### PR TITLE
tools/firmware-utils: refactor tp-link safeloader metadata partitions

### PR DESCRIPTION
**Pull request by svanheule**
In an attempt to bring a bit more structure to the tplink-safeloader image builder, I have started to refactor some of the code. Since this potentially touches a lot of code, I'm filing this as a draft PR, to check if this worth pursuing.

Currently, the changes are still split up, but e.g. the patches that touch the meta-partition padding can be merged. It's possible that the meta-partition padding values could be dropped altogether, but I don't have any devices where the stock upgrade images contain these padded partitions. I've noticed with my EAP-devices that the extra padding doesn't matter, since the amount of useful data is actually read from the meta-partition header.

To check whether the patches don't produce any (unintended) changes to the generated factory images, I've used a script that compares the (mock) output for all supported boards with a reference image built from master (ignoring the image's MD5 checksum).
After "tools/firmware-utils: safeloader extra-para is meta partition":
```
ARCHER-A7-V5 differs in 2 bytes
ARCHER-C7-V4 differs in 2 bytes
ARCHER-C7-V5 differs in 2 bytes
```
After "tools/firmware-utils: use same trailing data for all safeloader parts"
```
C2600 differs in 1 bytes
ARCHER-A7-V5 differs in 2 bytes
ARCHER-C5-V2 differs in 1 bytes
ARCHER-C7-V4 differs in 2 bytes
ARCHER-C7-V5 differs in 2 bytes
ARCHERC9 differs in 1 bytes
TLWA850REV2 differs in 1 bytes
TLWA855REV1 differs in 1 bytes
TLWR1043NDV4 differs in 1 bytes
TL-WR902AC-V1 differs in 1 bytes
TLWR942NV1 differs in 1 bytes
RE200-V2 differs in 1 bytes
RE220-V2 differs in 1 bytes
RE305-V1 differs in 1 bytes
RE350-V1 differs in 1 bytes
RE350K-V1 differs in 1 bytes
RE355 differs in 1 bytes
RE450 differs in 1 bytes
RE450-V2 differs in 1 bytes
RE450-V3 differs in 1 bytes
RE650-V1 differs in 1 bytes
```

The 2-byte changes come from the extra-para partition, while the 1-byte changes are caused by using the same padding value throughout the image.

PKG_RELEASE bumps will be added to the patches once I've got the number and order worked out.

### CODE REVIEW

**Comment by svanheule** [comment]
I've reworked the changes into two patches. A first focussing on the metadata partition generation, and a second on the padding of these metadata partitions. I haven't really been able to do more, and since this has been here for a while, I'll just leave it at these two patches for now.

**Comment by svanheule** [comment]
@adschm Would you be able to review these patches? I can also send them to the mailing list if you prefer that.

**Comment by adschm** [comment]
I prefer GitHub, but I can't review these (as goes for functional changes to tplink-safeloader in general, unless trivial).

**Comment by svanheule** [comment]
No problem. Let me ping a few more people here then, before trying the mailing list with the long patches.

@NeoRaider @jow- @lynxis You seem to have commited/authored some of the previous functional changes to tplink-safeloader. Anyone willing to review?

**Comment by lynxis** [comment]
I'll take a look later.

**Comment by svanheule** [comment]
Rebased onto master to also include TL-WPA8630P-V2.1-EU in "firmware-utils: tplink-safeloader: refactor meta-partition padding"

**Comment by svanheule** [comment]
Rebased commits to account for 65f3e7ce1f4a (PKG_RELEASE bump) and 33b76d66d1af (added TP-Link device).

File: tools/firmware-utils/src/tplink-safeloader.c [hunk 1]
```diff
@@ -2280,14 +2332,17 @@ static inline uint8_t bcd(uint8_t v) {
...
-	size_t part_len = sizeof(struct soft_version);
-	if (info->soft_ver_compat_level > 0)
-		part_len += sizeof(uint32_t);
-
-	struct image_partition_entry entry =
-	    alloc_image_partition("soft-version", part_len+1);
-	struct soft_version *s = (struct soft_version *)entry.data;
+static struct image_partition_entry make_soft_version(
+	const struct device_info *info, uint32_t rev
+) {
```
Comment (lynxis) [inline tools/firmware-utils/src/tplink-safeloader.c:2337]: can you move the ')' on the same line of the arguments.
The file already mixes different styles, but let's not introduce another style ;).

File: tools/firmware-utils/src/tplink-safeloader.c [hunk 2]
```diff
@@ -2215,6 +2229,44 @@ static inline void put32(uint8_t *buf, uint32_t val) {
...
+}
+
+/** Allocate a padded meta partition with a correctly initialised header
+ * If the `data` pointer is NULL, then the required space is only allocated,
+ * otherwise `data_len` bytes will be copied from `data` into the partition
+ * entry. */
+static struct image_partition_entry init_meta_partition_entry(
+	const char *name, const void *data, uint32_t data_len,
+	enum partition_trail_value pad_value
+) {
```
Comment (lynxis) [inline tools/firmware-utils/src/tplink-safeloader.c:2244]:
```
    enum partition_trail_value pad_value)
{
```

File: tools/firmware-utils/src/tplink-safeloader.c [hunk 3]
```diff
@@ -2298,58 +2353,51 @@ static struct image_partition_entry make_soft_version(struct device_info *info,
...
-	size_t len = strlen(info->support_list);
-	struct image_partition_entry entry = alloc_image_partition("support-list", len + 9);
-
-	put32(entry.data, len);
-	memset(entry.data+4, 0, 4);
-	memcpy(entry.data+8, info->support_list, len);
-	entry.data[len+8] = info->support_trail;
+static struct image_partition_entry make_support_list(
+	const struct device_info *info
+) {
```
Comment (lynxis) [inline tools/firmware-utils/src/tplink-safeloader.c:2387]: same

File: tools/firmware-utils/src/tplink-safeloader.c [hunk 4]
```diff
@@ -2298,58 +2353,51 @@ static struct image_partition_entry make_soft_version(struct device_info *info,
...
+	return init_meta_partition_entry(
+		"support-list", info->support_list, len, info->part_trail
+	);
+}
 
-	return entry;
+/** Partition with extra-para data */
+static struct image_partition_entry make_extra_para(
+	const struct device_info *info, const uint8_t *extra_para, size_t len
+) {
```
Comment (lynxis) [inline tools/firmware-utils/src/tplink-safeloader.c:2397]: same

**Review by lynxis — CHANGES_REQUESTED** [review]
Hi @svanheule,

only small nitpicks for coding style. But in general it looks good.
Thanks for the contributions.

Best,
lynxis

**Comment by lynxis** [comment]
It would be also great if you could rebase it as your patch for the EAP changed the same file.

**Comment by svanheule** [comment]
I've rebased for the added EAP devices, also dropping the partition padding there like for the ones that were already present.

The function definition line breaks have been reformatted. I've also reformatted some function calls, so a line break is only present if the next argument would exceed 80 characters.

**Comment by lynxis** [comment]
merged. thanks

**Pull request closed without merging**